### ISSUE TITLE
Removal of expired backups does not work

### ISSUE THREAD

**Issue by djablonski-moia** [vmware-tanzu/velero#3094]
**What steps did you take and what happened:**

In our AWS-based setup, when scheduled backup reach their TTL, the deletion process is started but gets stuck in status `Deleting`. The contents in S3 for the bucket are properly deleted, while volume snapshots stay (causing significant extra cost).

**What did you expect to happen:**

I expect backups to be cleanly removed when their TTL expires, including all backed up data, such as volume snapshots.

**The output of the following commands will help us better understand what's going on**:
(Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.)

- `kubectl logs deployment/velero -n velero`

```
time="2020-11-18T08:23:47Z" level=info msg="Removing existing deletion requests for backup" backup=velero-nightly-backup-20201103034355 controller=backup-deletion logSource="pkg/controller/backup_deletion_controller.go:469" name=velero-nightly-backup-20201103034355-gt6b9 namespace=velero
time="2020-11-18T08:23:50Z" level=error msg="Error in syncHandler, re-adding item to queue" controller=backup-deletion error="error downloading backup: error copying Backup to temp file: rpc error: code = Unknown desc = error getting object backups/velero-nightly-backup-20201103034355/velero-nightly-backup-20201103034355.tar.gz: NoSuchKey: The specified key does not exist.\n\tstatus code: 404, request id: 01DBEB5FABBF40BD, host id: HKe3B0heM0NpUhxXbLEZp7THCXtsfDKJkYdR6Sg0bS3+j0ywshitElmEnG7mPdDNmq6ASEtKT6w=" error.file="/go/src/github.com/vmware-tanzu/velero/pkg/controller/restore_controller.go:558" error.function=github.com/vmware-tanzu/velero/pkg/controller.downloadToTempFile key=velero/velero-nightly-backup-20201103034355-gt6b9 logSource="pkg/controller/generic_controller.go:140"
```

- `velero backup describe <backupname>` or `kubectl get backup/<backupname> -n velero -o yaml`

```
Name:         velero-nightly-backup-20201103034355
Namespace:    velero
Labels:       app.kubernetes.io/instance=velero
              app.kubernetes.io/managed-by=Tiller
              app.kubernetes.io/name=velero
              helm.sh/chart=velero-2.0.3
              velero.io/schedule-name=velero-nightly-backup
              velero.io/storage-location=aws
Annotations:  <none>

Phase:  Deleting

Errors:    0
Warnings:  0

Namespaces:
  Included:  *
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        <none>
  Cluster-scoped:  auto

Label selector:  <none>

Storage Location:  aws

Velero-Native Snapshot PVs:  auto

TTL:  360h0m0s

Hooks:  <none>

Backup Format Version:

Started:    2020-11-03 04:43:55 +0100 CET
Completed:  2020-11-03 04:52:17 +0100 CET

Expiration:  2020-11-18 04:43:55 +0100 CET

Velero-Native Snapshots:  2 of 2 snapshots completed successfully (specify --details for more information)

Deletion Attempts:
  2020-11-18 06:30:25 +0100 CET: InProgress
```

- `velero backup logs <backupname>`

```
Logs for backup "velero-nightly-backup-20201103034355" are not available until it's finished processing. Please wait until the backup has a phase of Completed or Failed and try again.
```

**Anything else you would like to add:**

My guess is that this at least loosely related to https://github.com/vmware-tanzu/velero/pull/2993.

**Environment:**

Velero version:
```
Client:
	Version: v1.5.2
	Git commit: -
Server:
	Version: v1.5.2
```

Velero features:
```
features: <NOT SET>
```

Kubernetes version:
1.18.9

Kubernetes installer & version:
kops 1.18.1

Cloud provider or hardware configuration:
AWS (with aws plugin)

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

**Comment by ashish-amarnath** [comment]
This is definitely a bug.
The correct fix is to not delete the Backup from the object store when there are errors.
In this case, there might have been errors deleting volumesnpshots at https://github.com/vmware-tanzu/velero/blob/main/pkg/controller/backup_deletion_controller.go#L349
And these errors are ignored to eventually delete the backup from the object store at https://github.com/vmware-tanzu/velero/blob/main/pkg/controller/backup_deletion_controller.go#L362

The correct fix should be to move the 
```
	if backupStore != nil {
		log.Info("Removing backup from backup storage")
		if err := backupStore.DeleteBackup(backup.Name); err != nil {
			errs = append(errs, err.Error())
		}
	}
```
into the
```
	if len(errs) == 0 {
		// Only try to delete the backup object from kube if everything preceding went smoothly
		err = c.backupClient.Backups(backup.Namespace).Delete(context.TODO(), backup.Name, metav1.DeleteOptions{})
		if err != nil {
			errs = append(errs, errors.Wrapf(err, "error deleting backup %s", kube.NamespaceAndName(backup)).Error())
		}
	}
```
at https://github.com/vmware-tanzu/velero/blob/main/pkg/controller/backup_deletion_controller.go#L410

cc @zubron because you are authoring PR #2993

**Comment by ashish-amarnath** [comment]
I am closing this as a duplicate of #2980
I've copied a link to my above comment into my code review of #2993

**Comment by ashish-amarnath** [comment]
https://github.com/vmware-tanzu/velero/pull/2993#issuecomment-730440597
Based on this comment reopening this and work on a fix for this.

**Comment by billimek** [comment]
I think that I'm experiencing this as well.  Is it 'safe' to manually delete the backups.velero.io objects that seem to be 'stuck' deleting (e.g. `k delete backups.velero.io -n velero velero-daily-backup-20201212060042`)

![image](https://user-images.githubusercontent.com/6393612/104255493-31cf1c00-5447-11eb-9930-9795b0188519.png)

**Comment by djablonski-moia** [comment]
@billimek For me it was the only way to clean this up, and I did not encounter any problems afterwards so far. But still, it's just guessing 😉

**Comment by kong62** [comment]
please add  ```velero backup delete --force``` parameter.
I do not think ```kubeclt delete backup``` is a good idea.

**Comment by reddiablo85** [comment]
any update on this?

**Comment by ashish-amarnath** [comment]
I am not working on this project anymore. Unassigning myself.

**Comment by wangxiaojin16** [comment]
> **What steps did you take and what happened:**
> 
> In our AWS-based setup, when scheduled backup reach their TTL, the deletion process is started but gets stuck in status `Deleting`. The contents in S3 for the bucket are properly deleted, while volume snapshots stay (causing significant extra cost).
> 
> **What did you expect to happen:**
> 
> I expect backups to be cleanly removed when their TTL expires, including all backed up data, such as volume snapshots.
> 
> **The output of the following commands will help us better understand what's going on**:
> (Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.)
> 
> * `kubectl logs deployment/velero -n velero`
> 
> ```
> time="2020-11-18T08:23:47Z" level=info msg="Removing existing deletion requests for backup" backup=velero-nightly-backup-20201103034355 controller=backup-deletion logSource="pkg/controller/backup_deletion_controller.go:469" name=velero-nightly-backup-20201103034355-gt6b9 namespace=velero
> time="2020-11-18T08:23:50Z" level=error msg="Error in syncHandler, re-adding item to queue" controller=backup-deletion error="error downloading backup: error copying Backup to temp file: rpc error: code = Unknown desc = error getting object backups/velero-nightly-backup-20201103034355/velero-nightly-backup-20201103034355.tar.gz: NoSuchKey: The specified key does not exist.\n\tstatus code: 404, request id: 01DBEB5FABBF40BD, host id: HKe3B0heM0NpUhxXbLEZp7THCXtsfDKJkYdR6Sg0bS3+j0ywshitElmEnG7mPdDNmq6ASEtKT6w=" error.file="/go/src/github.com/vmware-tanzu/velero/pkg/controller/restore_controller.go:558" error.function=github.com/vmware-tanzu/velero/pkg/controller.downloadToTempFile key=velero/velero-nightly-backup-20201103034355-gt6b9 logSource="pkg/controller/generic_controller.go:140"
> ```
> 
> * `velero backup describe <backupname>` or `kubectl get backup/<backupname> -n velero -o yaml`
> 
> ```
> Name:         velero-nightly-backup-20201103034355
> Namespace:    velero
> Labels:       app.kubernetes.io/instance=velero
>               app.kubernetes.io/managed-by=Tiller
>               app.kubernetes.io/name=velero
>               helm.sh/chart=velero-2.0.3
>               velero.io/schedule-name=velero-nightly-backup
>               velero.io/storage-location=aws
> Annotations:  <none>
> 
> Phase:  Deleting
> 
> Errors:    0
> Warnings:  0
> 
> Namespaces:
>   Included:  *
>   Excluded:  <none>
> 
> Resources:
>   Included:        *
>   Excluded:        <none>
>   Cluster-scoped:  auto
> 
> Label selector:  <none>
> 
> Storage Location:  aws
> 
> Velero-Native Snapshot PVs:  auto
> 
> TTL:  360h0m0s
> 
> Hooks:  <none>
> 
> Backup Format Version:
> 
> Started:    2020-11-03 04:43:55 +0100 CET
> Completed:  2020-11-03 04:52:17 +0100 CET
> 
> Expiration:  2020-11-18 04:43:55 +0100 CET
> 
> Velero-Native Snapshots:  2 of 2 snapshots completed successfully (specify --details for more information)
> 
> Deletion Attempts:
>   2020-11-18 06:30:25 +0100 CET: InProgress
> ```
> 
> * `velero backup logs <backupname>`
> 
> ```
> Logs for backup "velero-nightly-backup-20201103034355" are not available until it's finished processing. Please wait until the backup has a phase of Completed or Failed and try again.
> ```
> 
> **Anything else you would like to add:**
> 
> My guess is that this at least loosely related to #2993.
> 
> **Environment:**
> 
> Velero version:
> 
> ```
> Client:
> 	Version: v1.5.2
> 	Git commit: -
> Server:
> 	Version: v1.5.2
> ```
> 
> Velero features:
> 
> ```
> features: <NOT SET>
> ```
> 
> Kubernetes version:
> 1.18.9
> 
> Kubernetes installer & version:
> kops 1.18.1
> 
> Cloud provider or hardware configuration:
> AWS (with aws plugin)
> 
> **Vote on this issue!**
> 
> This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).
> Use the "reaction smiley face" up to the right of this comment to vote.
> 
> * 👍 for "I would like to see this bug fixed as soon as possible"
> * 👎 for "There are more important bugs to focus on right now"

Have you fixed it yet? I have the same issus too

**Comment by datle1** [comment]
> > **What steps did you take and what happened:**
> > In our AWS-based setup, when scheduled backup reach their TTL, the deletion process is started but gets stuck in status `Deleting`. The contents in S3 for the bucket are properly deleted, while volume snapshots stay (causing significant extra cost).
> > **What did you expect to happen:**
> > I expect backups to be cleanly removed when their TTL expires, including all backed up data, such as volume snapshots.
> > **The output of the following commands will help us better understand what's going on**:
> > (Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.)
> > 
> > * `kubectl logs deployment/velero -n velero`
> > 
> > ```
> > time="2020-11-18T08:23:47Z" level=info msg="Removing existing deletion requests for backup" backup=velero-nightly-backup-20201103034355 controller=backup-deletion logSource="pkg/controller/backup_deletion_controller.go:469" name=velero-nightly-backup-20201103034355-gt6b9 namespace=velero
> > time="2020-11-18T08:23:50Z" level=error msg="Error in syncHandler, re-adding item to queue" controller=backup-deletion error="error downloading backup: error copying Backup to temp file: rpc error: code = Unknown desc = error getting object backups/velero-nightly-backup-20201103034355/velero-nightly-backup-20201103034355.tar.gz: NoSuchKey: The specified key does not exist.\n\tstatus code: 404, request id: 01DBEB5FABBF40BD, host id: HKe3B0heM0NpUhxXbLEZp7THCXtsfDKJkYdR6Sg0bS3+j0ywshitElmEnG7mPdDNmq6ASEtKT6w=" error.file="/go/src/github.com/vmware-tanzu/velero/pkg/controller/restore_controller.go:558" error.function=github.com/vmware-tanzu/velero/pkg/controller.downloadToTempFile key=velero/velero-nightly-backup-20201103034355-gt6b9 logSource="pkg/controller/generic_controller.go:140"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * `velero backup describe <backupname>` or `kubectl get backup/<backupname> -n velero -o yaml`
> > 
> > ```
> > Name:         velero-nightly-backup-20201103034355
> > Namespace:    velero
> > Labels:       app.kubernetes.io/instance=velero
> >               app.kubernetes.io/managed-by=Tiller
> >               app.kubernetes.io/name=velero
> >               helm.sh/chart=velero-2.0.3
> >               velero.io/schedule-name=velero-nightly-backup
> >               velero.io/storage-location=aws
> > Annotations:  <none>
> > 
> > Phase:  Deleting
> > 
> > Errors:    0
> > Warnings:  0
> > 
> > Namespaces:
> >   Included:  *
> >   Excluded:  <none>
> > 
> > Resources:
> >   Included:        *
> >   Excluded:        <none>
> >   Cluster-scoped:  auto
> > 
> > Label selector:  <none>
> > 
> > Storage Location:  aws
> > 
> > Velero-Native Snapshot PVs:  auto
> > 
> > TTL:  360h0m0s
> > 
> > Hooks:  <none>
> > 
> > Backup Format Version:
> > 
> > Started:    2020-11-03 04:43:55 +0100 CET
> > Completed:  2020-11-03 04:52:17 +0100 CET
> > 
> > Expiration:  2020-11-18 04:43:55 +0100 CET
> > 
> > Velero-Native Snapshots:  2 of 2 snapshots completed successfully (specify --details for more information)
> > 
> > Deletion Attempts:
> >   2020-11-18 06:30:25 +0100 CET: InProgress
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * `velero backup logs <backupname>`
> > 
> > ```
> > Logs for backup "velero-nightly-backup-20201103034355" are not available until it's finished processing. Please wait until the backup has a phase of Completed or Failed and try again.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > **Anything else you would like to add:**
> > My guess is that this at least loosely related to #2993.
> > **Environment:**
> > Velero version:
> > ```
> > Client:
> > 	Version: v1.5.2
> > 	Git commit: -
> > Server:
> > 	Version: v1.5.2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Velero features:
> > ```
> > features: <NOT SET>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Kubernetes version:
> > 1.18.9
> > Kubernetes installer & version:
> > kops 1.18.1
> > Cloud provider or hardware configuration:
> > AWS (with aws plugin)
> > **Vote on this issue!**
> > This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).
> > Use the "reaction smiley face" up to the right of this comment to vote.
> > 
> > * +1 for "I would like to see this bug fixed as soon as possible"
> > * -1 for "There are more important bugs to focus on right now"
> 
> Have you fixed it yet? I have the same issus too

I also caught this issue. Do you resolve it ? I think @djablonski-moia should reopen this issue to find a complete solution.

**Comment by eleanor-millman** [comment]
Reopening this issue for now in case you or someone else wants to work on it. Unfortunately, the team is stretched too thinly right now to take this on, but we may do so in the future.

**Comment by eleanor-millman** [comment]
Additional explanation from a Velero dev that helped me understand this better: snapshots that aren't attached to a backup will hang around but not be accessible to a user, so the user is paying for their storage without any benefit.

**Comment by reasonerjt** [comment]
Based on the description it's not clear why the volume snapshot failed to be deleted.

After checking the code, it seems that the fix proposed in this [comment](https://github.com/vmware-tanzu/velero/issues/3094#issuecomment-730063746) is not needed.

Because with the fix in PR #2993 the error described in the description will be skipped and the backup_deletion_controller will try to delete the volume snapshot again, the deletion won't be blocked.

**Comment by reasonerjt** [comment]
@datle1 Because it involves deleting different resources, the reason deletion of a backup fails may vary.  Generally, I think it's correct behavior that the CR remains when any errors happen during the deletion.  The gc_controller will always try to delete the expired backup In the next cycle.
I'm closing this issue b/c I believe the error described in OP will not happen again.
Please upgrade to velero v1.7 or v1.8, and if you are still seeing errors when deleting backup please open another issue, I'll be happy to help you solve the problem.
